File: core/tests/fake_geometry_port_test.cpp
#include <gtest/gtest.h>

#include "toporoom/adapters/fake_geometry_port.hpp"
#include "toporoom/ports/geometry_port.hpp"

using toporoom::adapters::FakeGeometryPort;
using toporoom::ports::BuildRequest;
using toporoom::ports::FaultCode;
using toporoom::ports::FloorPlanSolidSemantics;
using toporoom::ports::semantics_from_scene_ir;
using toporoom::domain::SceneIR;
using toporoom::domain::SceneIRStorey;
using toporoom::domain::SceneIRWall;

TEST(FakeGeometryPort, ReturnsOkMeshWithWallHint) {
  SceneIR scene;
  scene.id = "doc_fake";
  scene.revision = 1;
  SceneIRStorey storey;
  storey.id = "storey_1";
  storey.height_mm = 2800;
  SceneIRWall wall;
  wall.id = "wall_s";
  wall.start = {0, 0};
  wall.end = {1000, 0};
  wall.thickness_mm = 200;
  wall.height_mm = 2800;
  storey.walls.push_back(wall);
  scene.storeys.push_back(storey);

  FakeGeometryPort port;
  BuildRequest request;
  request.document_rev = 1;
  request.rebuild_generation = 1;
  request.semantics = semantics_from_scene_ir(scene);
  const auto result = port.rebuild(request);
  ASSERT_TRUE(result.ok);
  ASSERT_FALSE(result.meshes.solids.empty());
  EXPECT_EQ(result.meshes.solids[0].node_hint, "Wall_wall_s");
}

TEST(FakeGeometryPort, ReturnsFaultWhenInstructed) {
  FakeGeometryPort port;
  port.fail_with({FaultCode::InvalidGeometry, "boom", {"wall_s"}});
  BuildRequest request;
  request.semantics.document_id = "doc_fake";
  const auto result = port.rebuild(request);
  EXPECT_FALSE(result.ok);
  EXPECT_EQ(result.fault.code, FaultCode::InvalidGeometry);
  EXPECT_EQ(result.fault.message, "boom");
}

File: packages/ports/src/fake-geometry-port.test.ts
import { describe, expect, it } from "vitest";
import { FakeGeometryPort, semanticsFromSceneIR } from "./fakes/index.js";

describe("FakeGeometryPort", () => {
  const semantics = semanticsFromSceneIR({
    format: "toporoom.sceneir",
    version: "0.1",
    id: "doc_fake",
    units: "mm",
    revision: 1,
    storeys: [
      {
        id: "storey_1",
        elevationMm: 0,
        heightMm: 2800,
        walls: [
          {
            id: "wall_s",
            kind: "exterior",
            start: { x: 0, y: 0 },
            end: { x: 1000, y: 0 },
            thicknessMm: 200,
            heightMm: 2800,
            openings: [],
          },
        ],
        rooms: [],
      },
    ],
    measurements: [],
  });

  it("returns an ok MeshProjection with Wall_ node hints", async () => {
    const port = new FakeGeometryPort();
    const result = await port.rebuild({
      documentRev: 1,
      rebuildGeneration: 1,
      dirty: true,
      semantics,
    });
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.meshes.solids[0]?.nodeHint).toBe("Wall_wall_s");
    }
  });

  it("returns a Fault when instructed", async () => {
    const port = new FakeGeometryPort();
    port.failWith({
      code: "InvalidGeometry",
      message: "boom",
      entityIds: ["wall_s"],
    });
    const result = await port.rebuild({
      documentRev: 1,
      rebuildGeneration: 1,
      dirty: true,
      semantics,
    });
    expect(result).toEqual({
      ok: false,
      fault: {
        code: "InvalidGeometry",
        message: "boom",
        entityIds: ["wall_s"],
      },
    });
  });
});
